docs/index.bs

```diff
@@ -2973,16 +2984,48 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
               1. For each |eventType| of |settingsObject|'s [=environment settings object/global object=]'s associated list of <a>event listeners</a>' event types:
                   1. [=set/Append=] |eventType| to |workerGlobalScope|'s associated [=ServiceWorkerGlobalScope/service worker=]'s <a>set of event types to handle</a>.
 
-                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's set of event types to handle remains an empty set. The user agents are encouraged to show a warning that the event listeners must be added on the very first evaluation of the worker script.
+                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's <a>set of event types to handle</a> remains an empty set.
 
               1. Set |script|'s <a>has ever been evaluated flag</a>.
+              1. If the [=Empty Handler identification=] algorithm with |workerGlobalScope| returns true, the [=all fetch listeners are empty flag=] is set.
+
+              Note: the user agents are encouraged to show warnings that the event listeners are no-op, and their executions can be skipped.
+
           1. Run the <a>responsible event loop</a> specified by |settingsObject| until it is destroyed.
           1. [=map/Clear=] |workerGlobalScope|'s [=map of active timers=].
       1. Wait for |serviceWorker| to be [=running=], or for |startFailed| to be true.
       1. If |startFailed| is true, then return *failure*.
       1. Return |serviceWorker|'s [=start status=].
   </section>
 
+  <section algorithm>
+    <h3 id="empty-handler-identification-algorithm"><dfn>Empty Handler identification</dfn></h3>
+
+      : Input
+      :: |workerGlobalScope|, a [=service worker/global object=].
+      : Output
+      :: a boolean
+
+      1. If |workerGlobalScope|'s <a>set of event types to handle</a> does not [=set/contain=] <code>fetch</code>, then returns false.
+      1. Let |eventListenerList| be an empty [=list=].
+      1. [=list/For each=] |eventListener| of |workerGlobalScope|'s <a>set of event types to handle</a>:
```

Comment:
This isn't right. `<a>set of event types to handle</a>` is a set of strings, not listeners.

In cases like this, where I want to get something like "all event listener callbacks of a particular type", I try to think of existing things that do that already that I could take inspiration from. In this case, I'd look at `target.dispatchEvent`, since it has to run all of the callbacks of a particular type.

So, I'd start at https://dom.spec.whatwg.org/#dom-eventtarget-dispatchevent, and follow the algorithm, and see how they gather up the callbacks. We probably need to do a similar thing here.

Following the trail takes me to https://dom.spec.whatwg.org/#concept-event-listener-invoke, which shows me how to get the listeners for a particular event target.

Then it goes to https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke, which shows me how to filter by event type.

Then it goes to https://webidl.spec.whatwg.org/#call-a-user-objects-operation which shows me how to get the callback from the two different forms of listener.

Reply:
Hmm, I got stuck on finding who sets currentTarget (https://dom.spec.whatwg.org/#dom-event-currenttarget) in Dispatching Events (https://dom.spec.whatwg.org/#concept-event-dispatch).
If that is cleared, we can get event listeners from currentTarget like invoke's step 6 (https://dom.spec.whatwg.org/#concept-event-listener-invoke), and the current algorithm can be used.
I tired to update anyway.